superpoint/models/super_point.py

```diff
@@ -0,0 +1,207 @@
+
+        im = inputs['image']
+        batch_size = config['batch_size']
+        shape = tf.shape(im)[1:3]
+
+        elems = tf.tile(tf.expand_dims(shape, 0), [batch_size, 1])
+        H = tf.map_fn(lambda s: sample_homography(s, **config['homographies']),
+                      elems, dtype=tf.float32)
+        H = tf.reshape(H, [batch_size, 8])
+        warped_im = tf.contrib.image.transform(im, H, interpolation="BILINEAR")
+        warped_im = tf.image.resize_images(warped_im, tf.floordiv(shape, 2))
+        if config['data_format'] == 'channels_first':
+            im = tf.transpose(im, [0, 3, 1, 2])
+            warped_im = tf.transpose(warped_im, [0, 3, 1, 2])
+
+        features = vgg_backbone(im, **config)
+        keypoints = detector_head(features, **config)
+        descriptors = descriptor_head(features, **config)
+
+        warped_features = vgg_backbone(warped_im, **config)
+        warped_keypoints = detector_head(warped_features, **config)
+        warped_descriptors = descriptor_head(warped_features, **config)
+
+        prob = keypoints['prob']
+        if config['nms']:
+            prob = tf.map_fn(lambda p: box_nms(p, config['nms'],
+                                               keep_top_k=config['top_k']), prob)
+            keypoints['prob_nms'] = prob
+        pred = tf.to_int32(tf.greater_equal(prob, config['detection_threshold']))
+        keypoints['pred'] = pred
+
+        prob = warped_keypoints['prob']
+        if config['nms']:
+            prob = tf.map_fn(lambda p: box_nms(p, config['nms'],
+                                               keep_top_k=config['top_k']),
+                             warped_keypoints['prob'])
+            warped_keypoints['prob_nms'] = prob
+        pred = tf.to_int32(tf.greater_equal(prob, config['detection_threshold']))
+        warped_keypoints['pred'] = pred
+
+        outputs = {'keypoints': keypoints,
+                   'descriptors': descriptors,
+                   'warped_keypoints': warped_keypoints,
+                   'warped_descriptors': warped_descriptors,
+                   'homography': H}
+
+        return outputs
+
+    def _loss(self, outputs, inputs, **config):
+        logits = outputs['keypoints']['logits']
+        warped_logits = outputs['warped_keypoints']['logits']
+        descriptors = outputs['descriptors']['logits']
+        warped_descriptors = outputs['warped_descriptors']['logits']
+
+        # Switch to 'channels last' once and for all
+        if config['data_format'] == 'channels_first':
+            logits = tf.transpose(logits, [0, 2, 3, 1])
+            warped_logits = tf.transpose(warped_logits, [0, 2, 3, 1])
+            descriptors = tf.transpose(descriptors, [0, 2, 3, 1])
+            warped_descriptors = tf.transpose(warped_descriptors, [0, 2, 3, 1])
+
+        # Resize the warped data to match the shape of the original data
+        shape = tf.shape(logits)[1:3]
+        warped_logits = tf.image.resize_images(warped_logits, shape)
```

Comment:
Wouldn't it be cleaner to instead rescale the homography fed to `warp_keypoints_to_map`?

Reply:
The issue is that the output of `warp_keypoints_to_map` is of the same shape as its input keypoint map, i.e. `inputs['keypoint_map']`, but we would like a ground truth twice as small if we don't resize `warped_logits` as it is the case currently.
One quick fix would be to modify `warp_keypoints_to_map` and add an argument setting the output shape that we want. What do you think?